discovery/pkg/sync/queue.go

```diff
@@ -103,38 +103,41 @@ func (sq *Queue) processNextWorkItem() bool {
 		return false
 	}
 
-	// We wrap this block in a func so we can defer sq.workqueue.Done.
-	err := func(obj interface{}) error {
-		// We call Done here so the workqueue knows we have finished processing
-		// this item. We also must remember to call Forget if we do not want
-		// this work item being re-queued. For example, we do not call Forget if
-		// a transient error occurs, instead the item is put back on the
-		// workqueue and attempted again after a back-off period.
-		defer sq.Workqueue.Done(obj)
-
-		action, ok := obj.(Action)
-		if !ok {
-			sq.Workqueue.Forget(obj)
-			sq.Metrics.QueueSizeGaugeMetric(sq.BackendName, sq.ClusterType, sq.Workqueue.Len())
-			return fmt.Errorf("ignoring unknown item of type %T in queue", obj)
-		}
-
-		err := action.Sync(sq.KubeClient, sq.Metrics, sq.BackendName)
-		if err != nil {
-			return err
-		}
-		sq.Logger.Infof("Successfully handled: %s", action)
+	// Tell the queue that we are done with processing this key. This unblocks
+	// the key for other workers.
+	defer sq.Workqueue.Done(obj)
 
-		// Finally, if no error occurs we Forget this item so it does not
-		// get queued again until another change happens.
+	action, ok := obj.(Action)
+	if !ok {
 		sq.Workqueue.Forget(obj)
 		sq.Metrics.QueueSizeGaugeMetric(sq.BackendName, sq.ClusterType, sq.Workqueue.Len())
-		return nil
-	}(obj)
+		sq.Logger.Errorf("got an unknown item of type %T in the queue", obj)
+		return true
+	}
+
+	err := action.Sync(sq.KubeClient, sq.Metrics, sq.BackendName)
 
-	if err != nil {
-		sq.Logger.Error(err)
+	// We successfully handled the action, so we can forget the item and keep going.
+	if err == nil {
+		sq.Workqueue.Forget(obj)
+		sq.Metrics.QueueSizeGaugeMetric(sq.BackendName, sq.ClusterType, sq.Workqueue.Len())
+		sq.Logger.Infof("Successfully handled: %s", action)
 		return true
 	}
+
+	// If there was an error handling the item, we will retry up to
+	// queueMaxRetries times.
+	if sq.Workqueue.NumRequeues(obj) < queueMaxRetries {
```

Comment:
I think we determined that `Forget()` didn't do much, but we still call it in the error-retry case but not the happy path.

Reply:
`Forget()` does have a function after all. The default rate limiter that we are using has two rate limiters. One of them is the `BucketRateLimiter`, for which `Forget()` has an empty implementation. The second is the `ItemExponentialFailureRateLimiter`, which does perform an operation when calling `Forget()`. 

My understanding is that `Forget()` does not have much to do with the queue itself, but with the rate limiters which are tracking each object in the queue, and determining when they are ready to be retried. By calling `Forget()`, we tell the rate limiter to stop tracking the object.

Thus, we call `Forget()` when we no longer want to retry an object. Or, on the flip side, the only place that we don't call `Forget()` is when we want the rate limiter to continue tracking the object because we are going to retry.

The updated (?) example in client-go talks a bit more about this: https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go

Reply:
👍